keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2262 @@
+  starts allocating it. Before it actually starts doing anything, the
+  driver adds a finalizer. Either adding the finalizer or removing the
+  ResourceClaim win. If the driver wins, it continues with allocation
+  and can either complete or abort the operation when it notices the non-nil
+  DeletionTimestamp. Otherwise, allocation gets aborted immediately.
+
+  What this avoids is the situation where an allocation succeed without having
+  an object where the result can be stored. The driver can also be killed at
+  any time: when it restarts, the finalizer indicates that allocation may be in
+  progress and has to be completed or aborted.
+
+  However, users may still force-delete a ResourceClaim, or the entire
+  cluster might get deleted. Driver implementations must store enough
+  information elsewhere to detect when some allocated resource is no
+  longer needed to recover from such scenarios.
+
+- In a cluster with multiple scheduler instances, two pods might get
+  scheduled concurrently by different schedulers. When they reference
+  the same ResourceClaim which may only get used by one pod at a time,
+  only one pod can be scheduled.
+
+  Both schedulers try to add their pod to the `ReservedFor` field, but only the
+  update that reaches the API server first gets stored. The other one fails
+  with a conflict error and the scheduler which issued it knows that it must
+  put the pod back into the queue, waiting for the ResourceClaim to become
+  usable again.
+
+### Custom parameters
+
+To support arbitrarily complex parameters, both ResourceClass and ResourceClaim
+contain one field which references a separate object. The reference contains
+API group, kind and name and thus is sufficient for generic clients to
+retrieve the parameters. For ResourceClass, that object must be
+cluster-scoped. For ResourceClaim, it must be in the same namespace as the
+ResourceClaim and thus the Pod. Which kind of objects a resource driver accepts as parameters depends on
+the driver.
+
+This approach was chosen because then validation of the parameters can be done
+with a CRD and that validation will work regardless of where the parameters
+are needed.
+
+A resource driver may support modification of the parameters while a resource
+is in use ("online resizing"). It may update the ResourceClaim status to
+reflect the modified state, for example by increasing the number of concurrent
+users. The driver must not allow the state to be modified such
+that a user of the resource no longer has access.
+
+Parameters may get deleted before the ResourceClaim or ResourceClass that
+references them. In that case, a pending resource cannot be allocated until the
+parameters get recreated. An allocated resource must remain usable and
+deallocating it must be possible. To support this, resource drivers must copy
+all relevant information:
+- For usage, the `AllocationResult.Attributes` can be hold the copied information
```

Comment:
This means converting an arbitrary params type to an opaque string (or somethign like that)?

Below you say "drivers should use some other location".  Wouldn't we be better to tell the driver to do this entirely in another type it owns?

* Admin creates Class GPU with driver gpu.example.com
* User creates a claim referencing class GPU and a gpu.example.com/GPUParams 
* Driver creates a gpu.example.com/GPUInstance with a sufficient copy of the class and claim params

We don't need to glom that into resourceClaim

To be clear - this is _a_ pattern.  They could just as easily store the info a a managed SQL database, if that's what they want - but it's THEIR problem, not ours.  We just need to tell them about it :)

Reply:
> This means converting an arbitrary params type to an opaque string (or somethign like that)?

It's a string->string map because that is common elsewhere (volume parameters), but basically yes.

> Below you say "drivers should use some other location". Wouldn't we be better to tell the driver to do this entirely in another type it owns?

We should encourage it, but I didn't want to make it a hard requirement for all drivers. Volumes are the same: a CSI driver must store information elsewhere to clean up orphaned volumes in all cases, but it also may store information in the container orchestrator's object (PersistentVolume in the case of Kubernetes).

For example, the example driver in the prototype just sets some env variables in the container. None of the user failure scenarios outlined here matter in that case, so just storing a copy of the intended env variables in `Attributes` is okay.

Reply:
> It's a string->string map because that is common elsewhere (volume parameters), but basically yes.

The history of string-string maps predates CRDs :)  Nowadays we use CRs for *everything* like this.

> We should encourage it, but I didn't want to make it a hard requirement for all drivers. 

It's not a hard requirement.  Thw requirement is "make sure that anything you allocate can later be deallocated, even if the end user does something pathological".  Practically that means "copy critical class data somewhere whose lifetime matches the claim's".  That could be a CR (with schema, validation, RBAC, etc) or it could be an annotation in the Claim (simpler, no schema, user can change it) or it could be something else entirely (e.g. something in the underlying resource's control plane).

Providing a string-string map here is only slightly better than an annotation, IMO.  Do we do this "save your opaque state here" in any other API?

Reply:
Yes, "VolumeHandle" and "VolumeAttributes" in [CSIPersistentVolumeSource](https://pkg.go.dev/k8s.io/api/core/v1#CSIPersistentVolumeSource), which is embedded inside a [PersistentVolume](https://pkg.go.dev/k8s.io/api/core/v1#PersistentVolume).

Reply:
I'm fine with replacing `Attributes` with a single string and let the driver handle encoding if they need anything more complex, but having some place that a driver can use if it suits them seemed reasonable.

Reply:
I mentioned elsewhere that storing any kind of tracking information for an allocation inside the cluster (no matter whether it is an annotation, this inline field, or a separate CR) cannot handle the situation where the cluster goes down and gets deleted.

Unfortunately I am not sure how a resource driver vendor would handle this case. First, the resource driver controller would have to be deployed outside of the cluster, otherwise it wouldn't be running and can never clean up. Second, it would need an API to distinguish "intermittent loss of connection to the API server, keep trying" from "cluster is gone, release resources". This sounds like something that a cloud provider can solve for drivers that they manage, but not a portable resource driver.

That then begs the question: is handling such a failure scenario part of this KEP? If not, where do we draw the line?

For CSI, the answers have been "the container orchestrator must store some state reliably" and "the user must not force-delete objects", otherwise volumes can leak. IMHO this is a reasonable compromise. Certain resource drivers may go beyond that and provide additional assurances, but not all of them need to (for example, a local resource will get freed automatically together with its node) and thus the API should support both kinds.

Reply:
> In CSI, the params are passed in a string map:

Yes, and was following that example in this KEP, but I agree that a totally opaque string is better. I just referenced it to show that allowing storing some data has been done before.

Let's make it a `ResourceHandle string` with a generous maximum size (16KiB?) and we'll see how this goes in practice.

> If your argument is "we don't need any of this, YOLO" then I am kind of with
you except I also see this as a UX thing - how can we show the user what
happened?

I think the burden falls on resource driver vendors here. They must document how they handle resource allocation, how users can query that (custom CRDs, out-of-cluster APIs), and what implications it has for cluster administration.

In the Kubernetes API description we can raise awareness of the problems, which is why I still think that mentioning the finalizer for ResourceClaim is useful. I've updated that part to cover additional failure scenarios.